### PR TITLE
Exclude taus, IVF and other high-level reco features from Run2_2018_pp_on_AA era

### PR DESCRIPTION
**Pull request by mandrenguyen**
Remove several components of the reconstruction that will not be used for heavy ion data (era Run2_2018_pp_on_AA):  The tau sequence is removed altogether.  It is re-added in PAT to keep it from crashing, although we will not write mini-AOD.  Inclusive vertexing, PF MET and, V0s and conversions are effectively removed by setting high thresholds.  This strategy keeps from having to fix a lot of crashes in mini-AOD, DQM, and validation that would arise due to missing collections.  These collections are rarely or never used in heavy ions and be created from AOD, which is the main output tier. 

These modifications, which leading to significant timing improvements,  were discussed at RECO/AT:
https://indico.cern.ch/event/711247/contributions/2920696/attachments/1614694/2565613/RecoTiming_180309_MAN.pdf

Tested with workflow 158.  
Will need to be backported to 10_2_X, if we use that for data-taking (there's no effect on pp reco)

### CODE REVIEW

**Comment by slava77** [comment]
@cmsbuild please test workflow 158.0

**Comment by andrius-k** [comment]
Hi Matthew,

I can't seem to find you in DQM Contacts list. Could you please introduce yourself briefly?

**Comment by mandrenguyen** [comment]
Hi @andrius-k , I'm heavy-ion software coordinator.  
For HIN DQM, our contact is @BetterWang

**Comment by andrius-k** [comment]
+1

File: RecoVertex/AdaptiveVertexFinder/python/inclusiveVertexing_cff.py [hunk 1]
```diff
@@ -59,7 +59,12 @@
...
+pp_on_XeXe_2017.toModify(inclusiveVertexFinder, minHits = 10, minPt = 1.0)
+pp_on_XeXe_2017.toModify(inclusiveCandidateVertexFinder, minHits = 10, minPt = 1.0)
+pp_on_XeXe_2017.toModify(inclusiveCandidateVertexFinderCvsL, minHits = 10, minPt = 1.0)
 from Configuration.Eras.Modifier_pp_on_AA_2018_cff import pp_on_AA_2018
-for e in [pp_on_XeXe_2017, pp_on_AA_2018]:
-    e.toModify(inclusiveVertexFinder, minHits = 10, minPt = 1.0)
-    e.toModify(inclusiveCandidateVertexFinderCvsL, minHits = 10, minPt = 1.0)
+pp_on_AA_2018.toModify(inclusiveVertexFinder, minHits = 999, minPt = 999.0)
+pp_on_AA_2018.toModify(inclusiveCandidateVertexFinder, minHits = 999, minPt = 999.0)
+pp_on_AA_2018.toModify(inclusiveCandidateVertexFinderCvsL, minHits = 999, minPt = 999.0)
```
Comment (slava77) [inline RecoVertex/AdaptiveVertexFinder/python/inclusiveVertexing_cff.py]: is there any use of b-tagging in HI?

Comment (mandrenguyen) [reply RecoVertex/AdaptiveVertexFinder/python/inclusiveVertexing_cff.py]: @slava77 there is some use of b-tagging, yes, but only for a few primary datasets.   For previous heavy-ion datasets we have run the SVing on AOD.  There's certainly a case to made for leaving this in, although it is quite time-consuming.  On the other hand, there is a high probability that we end up modifying the track selection for b-tagging (the performance is quite sensitive to this), and end up rerunning it anyway.   As it stands currently, the track selection is way too loose, as we are currently still using pp selections (@abaty can you confirm that?).   This leads to a ton of fakes in central events, which may be responsible for the long processing time.  Same goes for the PF linking, btw.  I propose to have another look when the MVA selections are in place, which will hopefully be soon.    If the timing is not too unreasonable, we can re-enable inclusiveCandidateVertexFinder.

Comment (slava77) [reply RecoVertex/AdaptiveVertexFinder/python/inclusiveVertexing_cff.py]: OK. I hope the future updates will keep things still relatively close to pp

**Comment by abaty** [comment]
>  as we are currently still using pp selections (abaty can you confirm that?)

This is correct, although we have started working on a new set of selections.

File: RecoTracker/ConversionSeedGenerators/python/ConversionStep_cff.py [hunk 2]
```diff
@@ -18,6 +18,10 @@
   trackClassifier       = cms.InputTag('tobTecStep',"QualityMasks"),
 )
 
+# turn off conversions for heavy ions without removing collections                                                                  
+from Configuration.Eras.Modifier_pp_on_AA_2018_cff import pp_on_AA_2018
+pp_on_AA_2018.toModify(convClusters, maxChi2 = 0.)
```
Comment (slava77) [inline RecoTracker/ConversionSeedGenerators/python/ConversionStep_cff.py]: it looks like this change doesn't do what the comment is implying.
IIUC, this maxChi2 is used to select clusters which are clearly belonging to a previous (tobtec) iteration tracks and should not be used in the following iteration.
By changing to maxchi2=0 all hits on tobtec tracks will be allowed to be considered in the conv step tracking.
So, the effect is kind of the opposite of what the comment says and perhaps this modification should be removed.

It looks like the following would achieve some reduction
``` photonConvTrajSeedFromSingleLeg.RegionFactoryPSet.RegionPSet.ptMin = 999.```

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
@cmsbuild please abort test

**Comment by slava77** [comment]
@cmsbuild please test workflow 158.0

**Comment by andrius-k** [comment]
+1

File: RecoTracker/ConversionSeedGenerators/python/ConversionStep_cff.py [hunk 3]
```diff
@@ -249,7 +250,8 @@
 #photonConvTrajSeedFromQuadruplets.primaryVerticesTag = cms.InputTag('pixelVertices')
 from Configuration.Eras.Modifier_trackingLowPU_cff import trackingLowPU
 trackingLowPU.toModify(photonConvTrajSeedFromSingleLeg, primaryVerticesTag   = "pixelVertices")
-
+from Configuration.Eras.Modifier_pp_on_AA_2018_cff import pp_on_AA_2018
+pp_on_AA_2018.toModify(photonConvTrajSeedFromSingleLeg.RegionFactoryPSet.RegionPSet, ptMin = 999.)
```
Comment (slava77) [inline RecoTracker/ConversionSeedGenerators/python/ConversionStep_cff.py]: sadly, this turned out to have even less effect.
I was just looking at parameters that look like a pt cut. Clearly, just having an apparently appropriate name is not good enough.

At a closer look, it seems like ```pp_on_AA_2018.toModify(photonConvTrajSeedFromSingleLeg, vtxMinDoF = 999999.)``` should actually do the job. This time I actually checked.

**Comment by slava77** [comment]
@cmsbuild please test workflow 158.0

**Comment by andrius-k** [comment]
+1

**Comment by slava77** [comment]
@mandrenguyen 
please remind me the rationale to turn off the convStepTracks (done via the seeding step now with the cut in photonConvTrajSeedFromSingleLeg).
Was it checked to be safe for the physics you'd be interested in?

In wf 158.0  these add up to ~50% of the allConversions collection.
These, in turn are used in egamma object IDs, including PF.
Based on 158.0 the effect on the PF or electron objects is rather small though (only a few PF photons are "lost"). 
This small effect may also be due to recently increased thresholds for egammas in pp_on_AA.

It could be that egamma objects, especially the less clean ones, which have conversions, are of little interest for HI.
Please clarify.
Thank you.

**Comment by mandrenguyen** [comment]
@slava77 Well, we didn't have conversion tracks in the old heavy-ion reco.  I'd be a bit worried about fakes in central events.  @ttrk @bi-ran @ikucher any opinion here?

**Comment by ttrk** [comment]
@mandrenguyen @slava77 
Island clustering does not use info about conversion tracks. GED photons contain whether there is an associated conversion track (https://github.com/cms-sw/cmssw/blob/master/RecoEgamma/EgammaPhotonProducers/src/GEDPhotonCoreProducer.cc#L102-L103). I made a check in MC. For pt > 20, fraction of GED photons with conversion tracks is <= 0.5%. Note that the fraction in peripheral (30-100%) events is ~0.5%, somewhat higher than in central (0-30%) events which is ~0.4%. Maybe centrality distorts association a bit.
In general, R9 is used as a proxy for conversion and it feeds into energy correction, not ID.
Let me know if you need more.

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/24224	1459953
- code changes are in line with the PR description and affect only workflows using the pp_on_AA_2018 modifier
- jenkins tests pass and comparisons with the baseline show no differences (the pp_on_AA_2018 is not tested)
- local tests with 230 "filtered on 3000" events and 200 wf 158.0 events show roughly expected behavior:
    - pfMet, pfChMet, inclusiveSecondaryVertices inclusiveCandidateSecondaryVertices, generalV0Candidates, and conversionStepTracks are now empty or zeroed out
    - the downstream effects are significant only in the combined b-taggers using the IVF and also in the "allConversions" with about 60% reduction in them. The "allConversions" play a role in the PF reco, but the net effect appears to be very small.
    - CPU time of the  "filtered on 3000" job in a production setup reduced from 240 to 160 s/event
    - output size of the AODSIM file ( "filtered on 3000" job) reduced by about 5%: about 3.6% in various vertices, 0.3% in V0s, and 0.45% in conversions
    - memory peak (RSS) for the 8-thread job ( "filtered on 3000") reduced from 15.8 GiB to 14.6 GiB    
        - this will be almost canceled out by the increase from #24257

**Comment by civanch** [comment]
+1

**Comment by kpedro88** [comment]
+1

**Comment by kpedro88** [comment]
merge

**Pull request merged**